crates/next-core/src/app_structure.rs

```diff
@@ -480,6 +480,18 @@ impl AppPageLoaderTree {
     }
 }
 
+#[turbo_tasks::value(transparent)]
+#[derive(Default)]
+pub struct RootParamVecOption(Option<Vec<RcStr>>);
```

Comment:
Is there a difference between `None` and `Some(vec![])`?

For me it looks like no. In this case I would change it to

```suggestion
pub struct RootParamVecOption(Vec<RcStr>);
```

Reply:
it's needed for the traversal in `directory_tree_to_entrypoints_internal_untraced`. when we encounter a root layout, we'll compute its root params based on the current `app_page` path. it should *not* be recomputed anywhere further below

so
- `None` signals "we don't know the root params for this subtree yet" (because we haven't found a root layout)
- `Some(vec!["foo"])` signals "we already have the root params, do not add anything more"
- `Some(vec![])` signals "this subtree has no root params" (which is possible if there's no dynamic segments that occur above a root layout)

for example, the Some/None distinction lets us avoid adding `bar` for page.tsx here:

```
app/
  [foo]/
    layout.tsx
    [bar]/
      page.tsx
```
- first, we'll visit `app/[foo]`, notice there's a layout (and `root_params` is `None`), and make `root_params = Some(["foo"])`.
- then, we'll visit app/[foo]/[bar]`, but we already have `root_params: Some(["foo"])` passed down from above, so we know not to compute them again. (which would be bad, because here we'd get `Some(["foo", "bar"])`)

it's similar for a case with no root params:
```
app/
  layout.tsx
  [bar]/
    page.tsx
```
we visit `app/`, find a root layout, and get `root_params = Some([])`, which'll also prevent us from making `bar` a root param.